botocore/serialize.py

```diff
@@ -600,30 +609,41 @@ def _partition_parameters(
         location = member.serialization.get('location')
         key_name = member.serialization.get('name', param_name)
         if location == 'uri':
-            partitioned['uri_path_kwargs'][key_name] = param_value
-        elif location == 'querystring':
-            if isinstance(param_value, dict):
-                partitioned['query_string_kwargs'].update(param_value)
-            elif isinstance(param_value, bool):
+            if isinstance(param_value, bool):
                 bool_str = str(param_value).lower()
-                partitioned['query_string_kwargs'][key_name] = bool_str
+                partitioned['uri_path_kwargs'][key_name] = bool_str
             elif member.type_name == 'timestamp':
                 timestamp_format = member.serialization.get(
-                    'timestampFormat', self.QUERY_STRING_TIMESTAMP_FORMAT
+                    'timestampFormat', self.URI_TIMESTAMP_FORMAT
                 )
                 timestamp = self._convert_timestamp_to_str(
                     param_value, timestamp_format
                 )
-                partitioned['query_string_kwargs'][key_name] = timestamp
+                partitioned['uri_path_kwargs'][key_name] = timestamp
             else:
-                partitioned['query_string_kwargs'][key_name] = param_value
+                partitioned['uri_path_kwargs'][key_name] = param_value
+        elif location == 'querystring':
+            if isinstance(param_value, dict):
+                # Add only new query string key/value pairs.
+                # Named query parameters should take precedence.
+                for key, value in param_value.items():
+                    partitioned['query_string_kwargs'].setdefault(key, value)
+            elif member.type_name == 'list':
+                new_param = [
+                    self._get_query_string_value(value, member.member)
+                    for value in param_value
+                ]
+                partitioned['query_string_kwargs'][key_name] = new_param
+            else:
+                new_param = self._get_query_string_value(param_value, member)
+                partitioned['query_string_kwargs'][key_name] = new_param
         elif location == 'header':
             shape = shape_members[param_name]
             if not param_value and shape.type_name == 'list':
                 # Empty lists should not be set on the headers
                 return
             value = self._convert_header_value(shape, param_value)
-            partitioned['headers'][key_name] = str(value)
```

Comment:
[Non-blocking, nit]: looks like you chose to do this because `_convert_header_value` will always return a string, right?  We could theoretically combine this to one line now without setting `value`, but it doesn't really make a difference:

`partitioned['headers'][key_name] = self._convert_header_value(shape, param_value)`